cl_manycore/libraries/bsg_manycore_cuda.cpp

```diff
@@ -47,312 +56,858 @@ static int hb_mc_get_y(tile_t *tiles, uint8_t *y_list, uint32_t num_tiles) {
+ * @param[in]  device        Pointer to device
+ * @parma[in]  grid_id       Id of grid to which the tile group belongs
+ * @param[in]  tg_id         Id of tile group
+ * @param[in]  grid_dim      X/Y dimensions of grid to which tile group belongs
+ * @parma[in]  dim           X/Y dimensions of tie pool (mesh)
+ * @param[in]  name          Kernel name that is to be executed on tile group
+ * @param[in]  argc          Number of input arguments to the kernel
+ * @param[in]  argv          List of input arguments to the kernel
  * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
  */
-int hb_mc_device_finish (uint8_t fd, eva_id_t eva_id, tile_t *tiles, uint32_t num_tiles) {
-	if (eva_id != 0) {
-		return HB_MC_FAIL; /* eva_id not supported */
-	} 
+int hb_mc_tile_group_enqueue (device_t* device, grid_id_t grid_id, hb_mc_coordinate_t tg_id, hb_mc_dimension_t grid_dim, hb_mc_dimension_t dim, char* name, uint32_t argc, uint32_t argv[]) {
 
-	else if (!mem_manager[eva_id])
-		return HB_MC_SUCCESS; /* there is no memory manager to deinitialize */
-	
-	delete(mem_manager[eva_id]);
+	if (device->num_tile_groups == device->tile_group_capacity) { 
+		device->tile_group_capacity *= 2;
+		device->tile_groups = (tile_group_t *) realloc (device->tile_groups, device->tile_group_capacity * sizeof(tile_group_t));
+		if (device->tile_groups == NULL) {
+			bsg_pr_err("%s: failed to allocate space for tile_group_t structs.\n", __func__);
+			return HB_MC_FAIL;
+		}
+	}
 	
-	for (int i = 0; i < num_tiles; i++) { /* freeze tiles */
-		hb_mc_tile_freeze_dep(fd, tiles[i].x, tiles[i].y);
+
+	tile_group_t* tg = (tile_group_t *)(device->tile_groups + device->num_tile_groups) ; 
+	tg->dim = dim;
+	tg->origin = device->mesh->origin;
+	tg->id = tg_id;
+	tg->grid_id = grid_id;
+	tg->grid_dim = grid_dim;
+	tg->status = HB_MC_TILE_GROUP_STATUS_INITIALIZED;
+
+	tg->map = (hb_mc_eva_map_t *) malloc (sizeof(hb_mc_eva_map_t)); 
+	if (tg->map == NULL) { 
+		bsg_pr_err ("%s: failed to allocate space for tile group hb_mc_eva_map_t struct map.\n", __func__);
+		return HB_MC_FAIL;
+	}
+	int error = hb_mc_origin_eva_map_init (tg->map, tg->origin); 
+	if (error != HB_MC_SUCCESS) { 
+		bsg_pr_err("%s: failed to initialize grid %d tile group (%d,%d) eva map origin.\n", __func__, tg->grid_id, tg->id.x, tg->id.y); 
+		return HB_MC_FAIL;
 	}
+	
 
-	int error = hb_mc_fifo_finish(fd);
-	if (error != HB_MC_SUCCESS) {
-		fprintf(stderr, "hb_mc_device_finish() --> hb_mc_host_finish(): failed to terminate host.\n");
+	tg->kernel = (kernel_t *) malloc (sizeof(kernel_t));
+	if (tg->kernel == NULL) { 
+		bsg_pr_err("%s: failed to allocated space for kernel_t struct.\n", __func__);
 		return HB_MC_FAIL;
 	}
+	tg->kernel->name = name;
+	tg->kernel->argc = argc;
+	tg->kernel->argv = argv;
+	tg->kernel->finish_signal_addr = FINISH_BASE_ADDR + ((tg->id.y * tg->grid_dim.x + tg->id.x) << 2); /* TODO: Hardcoded */
```

Comment:
I recommend making this a function. Maybe something like `tile_get_finish_signal_addr()`?

Reply:
Done, thanks!